README.md

```diff
@@ -1,16 +1,611 @@
+
+If serialized as JSON, the `sig` field MUST be serialized as [unpadded base64url](https://datatracker.ietf.org/doc/html/rfc4648#section-5).
+
+## 3.2 IPLD Schema
+
+``` ipldsch
+type Invocation struct {
+  inv &UCAN  -- The UCAN providing authority
+  v   SemVer -- Version
+  run Scope  -- Which actions to invoke
+  nnc String -- Nonce
+  ext Any    -- Extended fields
+  sig Bytes  -- Signature
+}
+
+type Scope enum {
+  | All ("*")
+  | Action
+}
+
+type Action enum {
+  using  URI 
+  do     Ability
+  inputs Any
+} 
+```
+
+## 3.3 JSON Examples
+
+### 3.3.1 Run All
+
+ ``` js
+{
+  "ucan/invoke": "QmYW8Z58V1v8R25USVPUuFHtU7nGouApdGTk3vRPXmVHPR",
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "ext": null,
+  "sig": "bdNVZn_uTrQ8bgq5LocO2y3gqIyuEtvYWRUH9YT-SRK6v_SX8bjt-VZ9JIPVTdxkWb6nhVKBt6JGpgnjABpOCA"
+}
+```
+
+ ``` js
+{
+  "ucan/invoke": "QmYW8Z58V1v8R25USVPUuFHtU7nGouApdGTk3vRPXmVHPR",
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "run": "*", // Exmplicitely "run all"
+  "ext": null,
+  "sig": "bdNVZn_uTrQ8bgq5LocO2y3gqIyuEtvYWRUH9YT-SRK6v_SX8bjt-VZ9JIPVTdxkWb6nhVKBt6JGpgnjABpOCA"
+}
+```
+
+### 3.3.2 Pipelines
+
+The following examples both express the following dataflow graph:
+
+```
+              ┌────────────────────────────┐
+              │                            │
+              │ dns://example.com?TYPE=TXT │
+              │        crud/update         │
+              │                            │
+              └───────┬──────────┬─────────┘
+                      │          │
+                      │          │
+┌─────────────────────▼───┐  ┌───▼────────────────────────┐
+│                         │  │                            │
+│ mailto:alice@exaple.com │  │ mailto://alice@example.com │
+│         msg/send        │  │          msg/send          │
+│     bob@example.com     │  │      carol@exmaple.com     │
+│                         │  │                            │
+└─────────────────────┬───┘  └───┬────────────────────────┘
+                      │          │
+                      │          │
+             ┌────────▼──────────▼────────┐
+             │                            │
+             │ https://example.com/events │
+             │         crud/create        │
+             │                            │
+             └────────────────────────────┘
+```
+
+#### 3.3.2.1 Batched 
+
+ ``` json
+{
+  "ucan/invoke": "QmYW8Z58V1v8R25USVPUuFHtU7nGouApdGTk3vRPXmVHPR",
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "ext": null,
+  "run": {
+    "update-dns" : {
+      "using": "dns://example.com?TYPE=TXT":
+      "do": "crud/update",
+      "inputs": { 
+         "value": "hello world",
+         "retries": 5
+      }
+    },
+    "notify-bob": {
+      "using": "mailto://alice@example.com",
+      "do": "msg/send",
+      "inputs": [
+        {
+          "to": "bob@example.com",
+          "subject": "DNSLink for example.com",
+          "body": {"ucan/promise": ["/", "update-dns", "$http.body"]}
```

Comment:
What’s $http.body here ?

Reply:
It's [JSONPath](https://www.npmjs.com/package/jsonpath), which has been around for a while and is [currently going through the IETF](https://datatracker.ietf.org/doc/draft-ietf-jsonpath-base/).

I was exploring what selector syntax to use because unlike IPVM we can't assume that someone is willing to load arbitrary user-defined Wasm to manage data (WAY more complexity). I was between a few syntaxes (including JSONPath and JSON Pointer), and @QuinnWilton vouched for JSONPath, which she's had successful for almost this exact use case in the past.

Reply:
I’m just not sure what “http” is referring to though, I don’t see “http” anywhere else, does it just gives a name to certain path ?

In other words I think it may be worth adding some explanations what happens in this example, because it is not necessarily clear for someone just reading through it

Reply:
Indeed! It's a path into the JSON _output_ of the function that hasn't returned yet

Reply:
And yes, adding more text there would be helpful. The example comes before the section on pipelines and receipts, so more commentary (or removing these features this early in the spec) would be helpful 💯 👍 Will do!

Reply:
I'm a fan of JSONPath, but it is worth keeping in mind that there isn't an RFC for it yet, like there is for JSON Pointer.

I think the choice between them really comes down to what sorts of selections you're looking to support though: JSON Pointer is great if you're looking to reference a single value in a JSON document, whereas JSONPath supports iteration + filtering, and can return multiple values as a result, while also performing lightweight computation using aggregates like `min` or `max`.

Where I think JSON Pointer really shines is alongside [JSON Reference](https://datatracker.ietf.org/doc/html/draft-pbryan-zyp-json-ref-03), which allows you to reference and transclude JSON values in other documents. This is used effectively in the OpenAPI specification to support both reusing definitions within a single file, or breaking up large specifications into multiple files: https://swagger.io/specification/#reference-object

JSONPath, on the other hand, ends up being useful if you're trying to compose JSON documents together in simple ways, without modifying the documents themselves. For example, in the context of UCAN invocation, it becomes possible to use a generic "HTTP Request" invocation that requests a JSON document, with the processing of the document happening using JSONPath. This means being able to represent common operations against the resulting JSON document using the invocation spec itself, such that precise meaning of the invocation is captured by the signature, without needing to define a comprehensive library of utility invocations to support those same operations.